typing/typecore.ml

```diff
@@ -859,7 +854,12 @@ let type_label_a_list ?labels loc closed env type_lbl_a opath lid_a_list k =
         (* Special case for rebuilt syntax trees *)
         List.map
           (function lid, a -> match lid.txt with
-            Longident.Lident s -> lid, Hashtbl.find labels s, a
+            Longident.Lident s ->
+              let lbl = Hashtbl.find labels s in
+              let lid =
+                Location.mkloc (Longident.Lident lbl.lbl_name) lid.loc in
+              lid,
+              Label.disambiguate () lid env opath (Ok[lbl, fun () -> ()]), a
```

Comment:
I still find this code somewhat dubious in the cases where we know the expected type (i.e. when opath is `Some ..`).
What happens in that case is:
- we know that the label we want here is `lbl` (that's what parmatch wants us to use)
- if `lbl` is a label for a type that matches the expected type, then use it
- otherwise, if the expected type has a label of the same name: use that

PS: the same remark applies for constructors.

Reply:
In the case we know the expected type, it will be used first.
There is just some extra code that can generate warnings, but they are disabled here.

It is only if there is no expected type that we use the received label. I kind of believe that this cannot happen, and that your approach would be sufficient, but I have no proof of that.

Reply:
That's not my point.

I think my point is more: if we're not going to completely trust disambiguation (and I agree that we shouldn't… yet), then I'd rather not use it at all. And I would just always use the received label.

It would just be simpler to think about.

Reply:
(Simpler than: "if this then that otherwise fallback with ...")

Reply:
As I explained above, in the current situation the unification might then lose information. We can change that after merging unique-types, then it would be completely equivalent.